Have DiscussionXBlock take care of loading assets it depends on

itsjeyd opened the pull request:
`DiscussionXBlock` currently relies on courseware to load JS and CSS assets that it depends on. This PR transfers responsibility for loading these assets to the XBlock itself.

cf. [YONK-382](https://openedx.atlassian.net/browse/YONK-382)

**Related PRs**

- https://github.com/edx-solutions/edx-platform/pull/744

**Screenshots**

- Adding a post: ![add-a-post](https://cloud.githubusercontent.com/assets/961441/18810662/83090878-829c-11e6-9f1a-469a22785fed.png)

- Expanded discussion: ![post-mathjax-expanded](https://cloud.githubusercontent.com/assets/961441/18810685/b89f427c-829c-11e6-890b-5f5da57cf9f0.png)

- Responses: ![response-comment-mathjax](https://cloud.githubusercontent.com/assets/961441/18810693/e67e05b6-829c-11e6-96eb-71fce703ddc2.png)

- Error message: ![post-with-error](https://cloud.githubusercontent.com/assets/961441/18810879/77f6cc2c-82a1-11e6-858a-e64b2fa8f5e8.png)

**Sandbox**

- LMS: http://pr13483.sandbox.opencraft.hosting/
- Studio: http://studio-pr13483.sandbox.opencraft.hosting/

Latest commit: 5b2a9d3caf5b9712d615b95495a9bef880c59fd9

<!--
b65a8fe0f89d506e7472ede5ec2eec7e4e297d54

7e400622e387c351c27b115151a298dda2f88cba

41be062be4c22c4e4a78e18865842a1cb20861be
-->

**Test instructions**

1. Remove static assets via `git clean -fx`.

2. Checkout this branch and rebuild assets via `paver update_assets` *or* `paver devstack studio`, `paver devstack lms`.

3. Add `DiscussionXBlock` to a unit via the main "Discussion" component button.

4. Navigate to the unit in the LMS and smoke test discussion functionality.

    For example, make sure that you can perform the following actions as `staff` user:

    * Create a Post, Reply, Comment.
    * Edit a Post, Reply, Comment.
    * Test [MathJax](http://docs.mathjax.org/en/latest/start.html) functionality in Post, Reply, Comment (and Preview), with e.g.
      ```
There are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$

When $a \ne 0$
      ```
    * Delete a Post, Reply, Comment.
    * Follow/unfollow a thread/post.
    * In the "Instructor" panel, add the `staff` user to the `Discussion Admin` or `Discussion Moderator` group. Then you should be able to: 
        * Open/close a thread/post.
        * Pin/unpin a thread/post.
    * Try to create a post without a title. This should cause an error message to be displayed above the controls you used to create the post.

5. Login as a regular student.  You should be able to do the above actions, plus:

    * Report/unreport a post created by the `staff` user.
    * Vote/unvote a thread/post created by the `staff` user.

6. Enable the "Teams" feature:

    * In Studio, replace the contents of "Advanced Settings" > "Teams Configuration" with:

      ```json
{
    "topics": [
        {
            "name": "Sustainability in Corporations",
            "description": "Description for Sustainability in Corporations",
            "id": "Sustain_Corporations"
        },
        {
            "name": "Water Conservation Projects",
            "description": "Description for Water Conservation",
            "id": "Water_Conservation"
        },
        {
            "name": "Sustainability Standards and Reporting",
            "description": "Description for Sustainability Standards",
            "id": "Standards_Reporting"
        }
    ],
    "max_team_size": 5
}
      ```
    * In the LMS, navigate to the "Teams" tab, select one of the topics and add a team to it via "create a new team in this topic".
    * Sanity-check discussion CSS by adding a post, response, and comment.

**Author notes**

* User profile page CSS looks broken, but is actually identical to what I get on `master`, so this is not a regression:

    * User profile page on `master`: ![user-profile-before](https://cloud.githubusercontent.com/assets/961441/18830693/205382c0-83e2-11e6-9e80-5491037f1e75.png)

    * User profile page on `itsjeyd/inline-discussion-assets`: ![user-profile-after](https://cloud.githubusercontent.com/assets/961441/18830700/29dadb4a-83e2-11e6-9321-eb1a69e26749.png)

<!--
* If you see an error such as "Uncaught ReferenceError: XBlock is not defined" from `runtime.v1.js` in the JS console, that means you have some stale assets lying around. Remove the following folders to get rid of the error:

    * `/static/coffee/src/xblock/`
    * `/static/coffee/src/discussion/`
-->

**Reviewers**

- [x] OpenCraft: @pomegranited 
- [ ] edX: @douglashall 
- [x] edX: @andy-armstrong

edx-webhook commented:
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=13483

e-kolpakov commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -108,6 +111,32 @@ def student_view(self, context=None):
 
+        # Head dependencies
+
+        discussion_vendor_js = ['js/jquery.autocomplete.js', 'js/src/tooltip_manager.js']
```
@itsjeyd nit: I would suggest extracting this into a method here or in dedicated module; when we get to course discussions (aka DiscussionCourseXBlock) we'll likely need to do something quite similar.

itsjeyd replied:
@e-kolpakov Yup, I was planning on doing that -- that's what I meant when I said the changes to `DiscussionXBlock` needed some cleaning up :)

Good news is, `DiscussionCourseXBlock` is already taking care of delivering the assets it depends on, see [xblock_discussion_course.py#L40](https://github.com/edx-solutions/edx-platform/blob/ziafazal/fix-broken-features/openedx/core/lib/xblock_builtin/xblock_discussion_course/xblock_discussion_course.py#L40), so there's a chance we won't have to update it again.

e-kolpakov commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -108,6 +111,32 @@ def student_view(self, context=None):
+
+        discussion_vendor_js = ['js/jquery.autocomplete.js', 'js/src/tooltip_manager.js']
+
+        discussion_vendor_js += settings.PIPELINE_JS['discussion_vendor']['source_filenames']
```
@itsjeyd I would expect it to just work, but have you checked if it still works in "pipeline" mode? I can't recall how exactly you switch to that - the idea is that pipeline is disabled in development env (default env for devstack) and all the files are just added one-by-one. In production, pipeline is enabled, so all those files are concatenated-minified-etc. so only one file per `PIPELINE_JS['entry']` is delivered. It would be nice to be sure that it still delivers one file when pipeline is enabled.

e-kolpakov commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -108,6 +111,32 @@ def student_view(self, context=None):
+
+        # Body dependencies
+
+        discussion_js = settings.PIPELINE_JS['discussion']['source_filenames']
```
@itsjeyd Same note as above about pipelines

e-kolpakov commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -108,6 +111,32 @@ def student_view(self, context=None):
+        for js_file in discussion_js:
+            fragment.add_javascript_url(static(js_file))
+
+        fragment.add_javascript(loader.render_template("static/js/discussion_classes.js"))
```
@itsjeyd `discussion_classes.js` is not a template - you might want to use `loader.load_unicode` (can't tell the exact name off top of my head, but this is likely accurate)

e-kolpakov commented on openedx/core/lib/xblock_builtin/xblock_discussion/static/js/discussion_classes.js:
```diff
@@ -0,0 +1,23 @@
+// Add RequireJS definitions for each discussion class
+
```
@itsjeyd Two suggestions:

1. This looks like a requireJS shim - might be good to reflect it in the file name.
2. Add TODO/task to refactor discussions JS into using RequireJS? Actually this might not be what we want - I'm not sure is there any support for RequireJS in Apros; if there are no such support this is not an option (at least until we add it).

itsjeyd replied:
@e-kolpakov Good point about RequireJS support in Apros, I'll have to check what the situation is.

e-kolpakov commented on lms/static/sass/inline-discussion.scss:
```diff
@@ -0,0 +1,33 @@
+
+// libs and resets *do not edit*
+@import 'bourbon/bourbon'; // lib - bourbon
+@import 'vendor/bi-app/bi-app-ltr'; // set the layout for left to right languages
```
@itsjeyd It uses LTR stylesheet, so

1. There should be a corresponding file with RTL + discussions should load the appropriate file (LTR vs RTL).
2. The file should have -ltr suffix, to stay in line with other direction-specific files.

itsjeyd replied:
@e-kolpakov I created a corresponding file for RTL, and updated `DiscussionXBlock` to conditionally load it (based on output of `get_language_bidi`).

I didn't add an `-ltr` suffix to the original file though: The naming convention for files in `lms/static/sass` is to add a suffix for RTL stylesheets, and to leave it off for LTR stylesheets (`lms-main-v1.scss` vs `lms-main-v1-rtl.scss`, etc.)

e-kolpakov commented on lms/static/sass/inline-discussion.scss:
```diff
@@ -0,0 +1,33 @@
+
+// discussion
+@import "course/discussion/form-wmd-toolbar";
+@import "course/discussion/form";
```
@itsjeyd not 100% sure about this, but is this `course/*` files for inline discussion or for course (aka tab) discussion?

itsjeyd replied:
@e-kolpakov Both of these files contribute to `lms-main.v1.css` on master, which gets loaded when accessing the "Course" tab. They are not included in `lms-discussion-main.css`, which gets loaded when accessing the "Discussion" tab.

e-kolpakov commented on lms/templates/discussion/_discussion_inline.html:
```diff
@@ -1,5 +1,8 @@
+
+<%include file="/mathjax_include.html" args="disable_fast_preview=disable_fast_preview"/>
 <%include file="_underscore_templates.html" />
+
```
@itsjeyd nit: unnecessary newline :)

itsjeyd replied:
@e-kolpakov I think I'll leave that in, the file as whole looks a little bit better with that extra bit of vertical whitespace (to me) :)

itsjeyd commented:
Thanks @e-kolpakov for the review! I addressed your comments; please have a look at the new commits. The only question I can't answer just yet is the [one](https://github.com/edx/edx-platform/pull/13483#r79129393) about RequireJS support in Apros.

e-kolpakov commented on lms/static/sass/inline-discussion-rtl.scss:
```diff
@@ -0,0 +1,34 @@
+// libs and resets *do not edit*
+@import 'bourbon/bourbon'; // lib - bourbon
+@import 'vendor/bi-app/bi-app-rtl'; // set the layout for right to left languages
+@import 'base/variables-rtl';
```
@itsjeyd nit: importing rtl variables before main variables seems a bit awkward - it won't allow overriding the var for RTL (if I understand how scss works right). Is it how it is done with other RTL files?

itsjeyd replied:
@e-kolpakov Good question. I based this file on [lms-main-v1-rtl.scss](https://github.com/edx/edx-platform/blob/master/lms/static/sass/lms-main-v1-rtl.scss), which also imports `base/variables-rtl` before `base/variables`. And it turns out that these files don't clash: `variables-rtl.scss` only defines a single variable (`$default-layout-direction`). The corresponding file for LTR is `variables-ltr.scss`, *but this isn't what gets imported in L9 below*. `_variables.scss` defines a bunch of essential variables; without these, this file does not compile. It doesn't touch `$default-layout-direction`, so it's fine to import it after `base/variables-rtl`.

e-kolpakov replied:
@itsjeyd ok - looks like there are two ways to properly handle this situation than:

1. RTL file should only contain direction var and should never clash with base var file. Inthis case it might be good to add a comment to *-rtl.* file(s).
2. RTL file is allowed to contain any styles and it should handle style overrides properly; in this case importing RTL before base is a bug that haven't exposed itself yet because RTL file does not contain any overrides.

Anyway, both are totally unrelated to this PR, so we can handle it separately, or just let it be until it becomes an actual bug (or is fixed occasionally in some future PR).

itsjeyd commented:
jenkins run lettuce

e-kolpakov commented:
jenkins don't listen to him :)

itsjeyd commented:
jenkins run js

itsjeyd commented:
jenkins run js

itsjeyd commented:
jenkins run lettuce

pomegranited commented:
@itsjeyd Found one more small thing to fix.  The issue existed prior to this PR, but if you can include it, that would be cool.

[forum-action-answer.underscore's `button.action-answer`](https://github.com/open-craft/edx-platform/blob/32ac77cca0dce5a3a1cebda18f2e979ef3a82f33/common/static/common/templates/discussion/forum-action-answer.underscore#L2) needs an additional class added: `btn-link`.

The issue can be seen when you Post a Question, and add a Reply.  The `Mark as Answer` button is misaligned with the other action buttons:

![screen shot 2016-09-26 at 11 52 49 am](https://cloud.githubusercontent.com/assets/7556571/18822869/71298968-83f3-11e6-8e0c-4b1dc8882384.png)

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js:
```diff
@@ -0,0 +1,35 @@
+    vendorScript = document.createElement('script');
+    vendorScript.onload = function() {
+        'use strict';
+        var MathJax, setMathJaxDisplayDivSettings;
```
This `MathJax` variable is obscuring `window.MathJax`.

Please change it to: `var MathJax = window.MathJax, setMathJaxDisplayDivSettings;`

pomegranited commented:
:+1: Once the [`MathJax` issue](https://github.com/edx/edx-platform/pull/13483#discussion_r80432205) is fixed.

- [x] I tested this by following the PR instructions on my devstack.
- [x] I read through the code.
- [x] I checked for accessibility issues by navigating through the PR instructions using only my keyboard.
- [x] <strike>Includes documentation</strike> code refactoring only; no documentation required.

itsjeyd commented:
jenkins run bokchoy

itsjeyd commented:
jenkins run lettuce

itsjeyd commented:
Thanks @pomegranited! Comments addressed. The user profile page looks the same on `master`, so this is not a regression (cf. updates to PR description).

@douglashall @andy-armstrong This is ready for you to have a look. See the PR description for sandbox links. Thanks!

itsjeyd commented:
jenkins run bokchoy

itsjeyd commented:
jenkins run lettuce

andy-armstrong commented:
@itsjeyd thanks for this PR. We'd been hoping to do this for a while now, so it is great to see it taken care of (see https://openedx.atlassian.net/browse/TNL-5302?jql=text%20~%20%22requirejs%20discussion%22). I'll review it in the next couple of days.

andy-armstrong commented:
@itsjeyd it looks like this has broken the styling in Teams, which is probably something you weren't aware of:

http://pr13483.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/teams/#teams/solar/andyas-team-bc897a2137014ede8e2bce9948df850f

![image](https://cloud.githubusercontent.com/assets/5985072/18836821/9dfe8ba6-83cf-11e6-92e6-5d79d2eebb0d.png)


The three places that discussions appear today are inline in the course, on the discussion tab (we call this variant the discussion board), and in teams.

itsjeyd commented:
Thanks @andy-armstrong! I'll look into fixing the CSS for the Teams feature.

itsjeyd commented:
@andy-armstrong The latest commits fix the CSS regressions for the Teams feature. I updated the sandbox, so you can verify the fix there.

CC @pomegranited

andy-armstrong commented on lms/static/sass/inline-discussion.scss:
```diff
@@ -0,0 +1,22 @@
+// From lms-main-v1.scss
```
Same question as from the rtl file: why should all these imports be duplicated? I'd rather see a shared Sass partial that both files can include.

itsjeyd replied:
@andy-armstrong This commit addresses your comment: https://github.com/edx/edx-platform/pull/13483/commits/524599ea2e469fc90a14ec71ac29c4d7a99ad328.

andy-armstrong replied:
Thanks. I'd remove the comment now, and replace it with a header that says what this file is.

itsjeyd replied:
@andy-armstrong Done, see https://github.com/edx/edx-platform/pull/13483/commits/e811df70457d6ece87f51d76319fd89cbce6c456.

andy-armstrong commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -108,6 +157,8 @@ def student_view(self, context=None):
 
+        self.add_resource_urls(fragment)
+
         course = self.runtime.modulestore.get_course(self.course_key)
```
This wasn't your change, but it seems expensive to have to get the whole course just to render this one XBlock. In looking at the template, it doesn't seem like the course context variable is used, but it is hard to tell due to template inheritance. Could you try removing it and see if it is necessary. Hopefully we don't force XBlocks to have to know how to do this.

itsjeyd replied:
@andy-armstrong Doesn't look like this is needed. I removed it and was able to interact with `DiscussionXBlock` instances as expected. See https://github.com/edx/edx-platform/pull/13483/commits/1abfac75b33cad44d24aceb65c95c67fb4f2bf34.

andy-armstrong commented on lms/templates/courseware/courseware.html:
```diff
@@ -61,7 +61,10 @@
 <%static:css group='style-student-notes'/>
 % endif
 
-<%include file="../discussion/_js_head_dependencies.html" />
```
Just FYI, courseware_chromeless is used to render XBlocks on mobile. You can manually see how this works by hitting this URL:

http://pr13483.sandbox.opencraft.hosting/xblock/block-v1:edX+DemoX+Demo_Course+type@vertical+block@vertical_3888db0bc286

andy-armstrong replied:
Of course, discussions are rendered natively on mobile, so this may not be necessary. However, it is good to keep the two files in sync, and in fact it will be much clearer if the chromeless version no longer includes discussion files.

itsjeyd replied:
@andy-armstrong Thanks for pointing this out. Made the corresponding change via https://github.com/edx/edx-platform/pull/13483/commits/b8f9caf7c95d9ed293178ddd27f16aabd8a3e6fc.

itsjeyd commented:
Thank you @andy-armstrong for the review. I'll have a closer look at your comments soon.

itsjeyd commented on lms/templates/courseware/courseware.html:
```diff
@@ -73,7 +76,7 @@
 
   <%static:js group='courseware'/>
-  <%include file="../discussion/_js_body_dependencies.html" />
+  <%include file="/mathjax_include.html" args="disable_fast_preview=disable_fast_preview"/>
```
@andy-armstrong It's not just discussions that need MathJax. For example, the "Numerical Input with Hints and Feedback" and "Math Expression Input" problem types need it too. So I don't think we can remove it from the "Course" tab.

andy-armstrong replied:
Thanks for clarifying. The code is also now much clearer in that it makes the MathJax dependency explicit, rather than having it hidden in a discussion file. I guess I have to wait until all xmodules are converted to XBlocks before we can stop loading MathJax on every page.

itsjeyd commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -88,6 +94,49 @@ def django_user(self):
+            fragment.add_css_url(static(css_file))
+
+        # Body dependencies
+        fragment.add_javascript(loader.load_unicode("static/js/mathjax_include.js"))
```
@andy-armstrong A quick question, to make sure I understand correctly: What exactly are you referring to when you say "discussion JS bundle"? :)

andy-armstrong replied:
@itsjeyd sorry that wasn't clear! I was referring to the files that all get bundled into discussion.js via this rule:

https://github.com/edx/edx-platform/blob/master/lms/envs/common.py#L1575

What I want to ensure is that in production we don't include extra JavaScript in the rendered HTML for the block, but at the same time we don't introduce an extra file download. The ideal situation to my mind is that everything goes into discussion.js.

itsjeyd commented on lms/static/sass/_build-course.scss:
```diff
@@ -60,8 +60,7 @@
 // discussion
-@import "course/discussion/form-wmd-toolbar";
-@import "course/discussion/form";
+@import 'build-discussion-editor';
```
@andy-armstrong I looked into this. Here is what I found:

* `course/discussion/form` targets `.wmd-prompt-dialog`
* `.wmd-prompt-dialog` comes from `Markdown.Editor.js`
* `Markdown.Editor.js` is only referenced in `discussion_vendor_js`
* `course/discussion/form-wmd-toolbar` only targets elements that belong to discussion-related components

Based on this,  `course/discussion/form-wmd-toolbar` and `course/discussion/form` are not needed here. We'll need to make sure they are loaded on the "Teams" tab though; the template for that tab [imports](https://github.com/open-craft/edx-platform/blob/itsjeyd/inline-discussion-assets/lms/djangoapps/teams/templates/teams/teams.html#L17) `lms-course.css`, so it's relying on the styles defined here. The approach you describe [here](https://github.com/edx/edx-platform/pull/13483/files#r80835975) should work.

itsjeyd commented on lms/static/sass/_build-lms-v1.scss:
```diff
@@ -62,20 +62,7 @@
-@import "discussion/views/response";
-@import 'discussion/utilities/developer';
-@import 'discussion/utilities/shame';
+@import 'build-discussion';
```
@andy-armstrong Yes, this was here to make the team's styling work. I had removed discussion-related code from this file in a previous iteration to keep the rules from getting included twice, but that was before you pointed out the "Teams" feature :)

Implemented your suggestion in https://github.com/edx/edx-platform/pull/13483/commits/b50a7dacf9b2da92e832b50f90500a5cc94eccdf.

itsjeyd commented on lms/static/sass/inline-discussion-rtl.scss:
```diff
@@ -0,0 +1,23 @@
+// From lms-main-v1-rtl.scss
```
@andy-armstrong Yup, that makes sense. This commit addresses your comment: https://github.com/edx/edx-platform/pull/13483/commits/524599ea2e469fc90a14ec71ac29c4d7a99ad328.

andy-armstrong replied:
Thanks. I'd remove the comment now, and replace it with a header that says what this file is.

itsjeyd replied:
@andy-armstrong Done, see https://github.com/edx/edx-platform/pull/13483/commits/e811df70457d6ece87f51d76319fd89cbce6c456.

itsjeyd commented on lms/static/sass/_build-discussion.scss:
```diff
@@ -0,0 +1,16 @@
+// From _build-lms-v1.scss
```
@andy-armstrong Thanks for the info. This is a refactoring I'd rather tackle in a separate PR: It's not a requirement for the current goal of having `DiscussionXBlock` take care of loading assets it depends on (as specified by [YONK-382](https://openedx.atlassian.net/browse/YONK-382)).

andy-armstrong replied:
Fair enough. Could you just move the files into the discussion subdirectory, then. That will keep them more clearly separated. Thanks.

itsjeyd replied:
This particular thread of comments continues [here](https://github.com/edx/edx-platform/pull/13483#pullrequestreview-1985827).

itsjeyd commented:
@andy-armstrong Thanks a lot for the thorough review.

Here is a status update: I addressed all of your comments regarding Sass, and will continue with the remaining comments tomorrow. If you could have a look at the changes I pushed today in the meantime, that would be great -- if you have any follow-up comments, I'd be able to address them tomorrow morning (along with the other comments). I'm in UTC+2, so that would give me a chance to get everything ready by the time you start your day.

(I'll also need to apply any additional changes I make here to https://github.com/edx-solutions/edx-platform/pull/744 to keep the differences minimal.)

It would be great to get this merged this week. If that doesn't happen, I'll have to hand it over to someone else from my team: I'll be out-of-office for a week starting Monday, so someone else would have to take over to make sure these changes get merged before the next sprint deadline. It would be great to be able to avoid that :)

andy-armstrong commented on lms/static/sass/_build-discussion-editor.scss:
```diff
@@ -0,0 +1,4 @@
+// From _build-course.scss
+
+@import "course/discussion/form-wmd-toolbar";
```
Can this just be moved into _build-discussion.scss? It seems to always be used together.

itsjeyd replied:
@andy-armstrong Yes, that works. See https://github.com/edx/edx-platform/pull/13483/commits/e811df70457d6ece87f51d76319fd89cbce6c456.

andy-armstrong commented on openedx/core/lib/xblock_builtin/xblock_discussion/static/js/discussion_classes.js:
```diff
@@ -0,0 +1,22 @@
+// Add RequireJS definitions for each discussion class
```
I still have these two questions.

itsjeyd replied:
@andy-armstrong I tested without this shim, and things worked fine. Removed it in https://github.com/edx/edx-platform/pull/13483/commits/724a839f77255346e911b8de2468d8ec5f4c68ea.

andy-armstrong commented on openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js:
```diff
@@ -0,0 +1,36 @@
+var vendorScript;
```
I'd like to see these changes made.

itsjeyd replied:
@andy-armstrong I changed the code to include this file in the JS discussion bundle: https://github.com/edx/edx-platform/pull/13483/commits/7d311c378542ad17b7676c10fa2834a6b307ee28.

As for refactoring `mathjax_include.html`: This is tricky because it is used on several different pages (and not just for discussion posts), with varying values of `disable_fast_preview` and `mathjax_mode`:

-   Wiki
    -   lms/templates/wiki/base.html
        -   `disable_fast_preview`: True
        -   `mathjax_mode`: "wiki"
    -   lms/templates/wiki/preview_inline.html
        -   `disable_fast_preview`: True
        -   `mathjax_mode`: "wiki"

-   Teams (via _js_body_dependencies.html)
    -   lms/djangoapps/teams/templates/teams/teams.html
        -   `disable_fast_preview`: True
        -   `mathjax_mode`: Undefined

-   Discussion (via _js_body_dependencies.html)
    -   lms/djangoapps/discussion/templates/discussion/discussion_profile_page.html
        -   `disable_fast_preview`: True
        -   `mathjax_mode`: Undefined
    -   lms/djangoapps/discussion/templates/discussion/discussion_board.html
        -   `disable_fast_preview`: False
        -   `mathjax_mode`: Undefined

-   Course
    -   lms/templates/courseware/courseware.html
        -   `disable_fast_preview`: False
        -   `mathjax_mode`: Undefined
    -   lms/templates/courseware/courseware-chromeless.html
        -   `disable_fast_preview`: False
        -   `mathjax_mode`: Undefined

-   "Static tab"
    -   lms/templates/courseware/static_tab.html
        -   `disable_fast_preview`: True
        -   `mathjax_mode`: Undefined

So what I did to create this file (`mathjax_include.js`) was to extract the (non-contiguous) regions of code that applied to the "Course" tab from `mathjax_include.html`.

To completely DRY out `mathjax_include.html`, it seems that we'd need to create multiple JS files from the different sections of that file, and then include them based on the values of `disable_fast_preview` and `mathjax_mode`. An alternative approach would be to remove `mathjax_include.html` and include the individual JS files directly in each of the templates (again based on the values of `disable_fast_preview` and `mathjax_mode`). In any case, the code in the individual JS files would need to be reusable so that this file (`mathjax_include.js`) could combine it as it does now.

Since this would impact many important parts of the system, and goes beyond enabling `DiscussionXBlock` to load assets it depends on, I think it would make sense to handle it in a separate PR.

andy-armstrong replied:
@itsjeyd Thanks for the detailed explanation. 👍  with your solution and we'll have to revisit this later.

andy-armstrong commented on openedx/core/lib/xblock_builtin/xblock_discussion/utils.py:
```diff
@@ -0,0 +1,17 @@
+"""
+Utils for DiscussionXBlock
```
I still think this should be moved.

itsjeyd replied:
@andy-armstrong I forgot to ask last night: What would be a good place to move this to in your opinion? It seems like it could probably live somewhere in `common`, since it could be useful to both Studio and LMS code. However, none of the top-level directories under `common` seem like a good fit. I thought about moving the file to `common/djangoapps/pipeline_js`; but now that the file also defines a function for obtaining CSS dependencies (via https://github.com/edx/edx-platform/pull/13483/commits/2877d907fb27b77f749d20866a9a4f05c65f6389), I'm not sure if `common/djangoapps/pipeline_js` is the best fit.

andy-armstrong replied:
Hmm, this is tricky. I agree that pipeline_js seems like a reasonable home, except that it is xmodule-centric and CSS methods don't fit naturally. The only other place that uses ``PIPELINE_ENABLED`` is ``openedx/core/lib/xblock_utils.py`` which is at least not discussion specific. Maybe put them there and at some point in the future we could have a cleaner pipeline utils location.

itsjeyd replied:
@andy-armstrong OK, sounds good. Made the change in https://github.com/edx/edx-platform/pull/13483/commits/ad5f69a9bcd408962eb54f32ab83fdd18394d7d1, and added some tests for the functions in https://github.com/edx/edx-platform/pull/13483/commits/b65a8fe0f89d506e7472ede5ec2eec7e4e297d54.

andy-armstrong commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -88,6 +94,49 @@ def django_user(self):
+    @staticmethod
+    def vendor_js_dependencies():
+        """
+        Returns list of vendor JS files that this XBlock depends on.
```
I'd still like more detailed docstring for these functions.

itsjeyd replied:
@andy-armstrong Done, see https://github.com/edx/edx-platform/pull/13483/commits/2877d907fb27b77f749d20866a9a4f05c65f6389.

andy-armstrong commented:
@itsjeyd thanks for addressing all the issues with the Sass. I've reviewed all your changes and while I have a couple of minor comments, this is overall looking great. I'll do another review tomorrow when you finish addressing the rest of my comments, and I'll make sure to be prompt so that we can get the PR merged this week.

itsjeyd commented on openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py:
```diff
@@ -88,6 +94,49 @@ def django_user(self):
+        for js_file in self.js_dependencies():
+            fragment.add_javascript_url(static(js_file))
+
+        fragment.add_javascript(loader.load_unicode("static/js/discussion_classes.js"))
```
@andy-armstrong Removed in https://github.com/edx/edx-platform/pull/13483/commits/724a839f77255346e911b8de2468d8ec5f4c68ea.

itsjeyd commented:
jenkins run js

itsjeyd commented:
jenkins run js

itsjeyd commented:
Thank you @andy-armstrong for the additional feedback. I addressed your comments. Please have another look.

itsjeyd commented:
Rebased from https://github.com/edx/edx-platform/pull/13483/commits/1b4058f423397d107f732a4a9c30356e01445ff0.

itsjeyd commented:
jenkins run js

pomegranited commented:
Rebased from 5b2a9d3.

pomegranited commented:
jenkins run js

pomegranited commented:
jenkins run lettuce

pomegranited commented:
Hi @andy-armstrong !  @itsjeyd is on holiday, so I'm helping finish up this PR.

But I'm really puzzled by the test failures.  I've asked my team for help, but maybe you how to fix them?

* [jenkins/js](https://build.testeng.edx.org/job/edx-platform-js-pr/25781/) is showing failure, and yet reports "[Test Result](https://build.testeng.edx.org/job/edx-platform-js-pr/25781/testReport/) (no failures)".  Is this a worry?
* [jenkins/python](https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/26002/) reports 12 failures that look totally unrelated to this change.
    * I can reproduce 10/12 of them by running `paver test_lib -l common/lib/modulestore/` on my devstack.
    * It looks like they might all be caused by [this added include](https://github.com/edx/edx-platform/pull/13483/files#diff-e3cb55b56190862761d9641bffe27ac9R19), which introduces a dependency on [`common/djangoapps/static_replace`](https://github.com/edx/edx-platform/tree/master/common/djangoapps/static_replace).
    * Traceback:

        ```
ERROR: test_publish_draft_delete (xmodule.modulestore.tests.test_publish.TestPublish)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/tests/test_split_w_old_mongo.py", line 69, in setUp
    self._create_course()
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/tests/test_publish.py", line 81, in _create_course
    split=False
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/tests/test_split_w_old_mongo.py", line 87, in _create_item
    runtime=self.runtime
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/mongo/base.py", line 1402, in create_item
    xblock = self.create_xblock(runtime, course_key, block_type, block_id, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/mongo/draft.py", line 323, in create_xblock
    runtime, course_key, block_type, block_id, fields, **kwargs
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/mongo/base.py", line 1362, in create_xblock
    xblock_class = runtime.load_block_type(block_type)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1444, in load_block_type
    return super(DescriptorSystem, self).load_block_type(block_type)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 606, in load_block_type
    return XBlock.load_class(block_type, self.default_class, self.select)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/plugin.py", line 118, in load_class
    PLUGIN_CACHE[key] = cls._load_class_entry_point(selected_entry_point)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/plugin.py", line 73, in _load_class_entry_point
    class_ = entry_point.load()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2355, in load
    return self.resolve()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2361, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/edx/app/edxapp/edx-platform/openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion.py", line 19, in <module>
    from openedx.core.lib.xblock_utils import get_css_dependencies, get_js_dependencies
  File "/edx/app/edxapp/edx-platform/openedx/core/lib/xblock_utils/__init__.py", line 10, in <module>
    import static_replace
ImportError: No module named static_replace
        ```

andy-armstrong commented:
@pomegranited The JavaScript tests failed with an error outside of a particular test, which is why the report is confusing. If you look at the full console log, you can see that there is an issue with MathJax:

```
08:45:55 Firefox 42.0.0 (Linux 0.0.0) ERROR
08:45:55   Script error.
08:45:55   at https://cdn.mathjax.org/mathjax/2.6-latest/config/TeX-MML-AM_SVG.js?rev=2.6.1
08:45:55 Firefox 42.0.0 (Linux 0.0.0) ERROR
08:45:55   Script error.
08:45:55   at https://cdn.mathjax.org/mathjax/2.6-latest/config/TeX-MML-AM_SVG.js?rev=2.6.1
08:45:55 
08:45:55 Firefox 42.0.0 (Linux 0.0.0): Executed 107 of 282 ERROR (3.721 secs / 3.091 secs)
```

andy-armstrong commented:
jenkins run js

andy-armstrong commented:
jenkins run python

andy-armstrong commented:
@pomegranited The Python failures seem to be that Django classes are now accidentally being loaded during the non-Django tests. In particular, the file ``openedx/core/lib/xblock_utils/__init__.py`` imports a ton of Django classes which seems like a big no-no to me, given that this file is not beneath the ``openedx/djangoapps`` directory.

I think this was caused by moving two helper functions into this file. I'd move them somewhere else where they won't accidentally cause Django imports to happen.

andy-armstrong commented:
@pomegranited it looks like that file also caused a merge conflict, so hopefully that will resolve matters. I'd do a rebase anyway just to be sure.

pomegranited commented:
Rebased from 105384e.

pomegranited commented:
jenkins run bokchoy

pomegranited commented:
Thanks for your advice, @andy-armstrong !

* python tests: 29e0674 moves those new functions into `openedx.core.lib.xblock_builtin`, and hopefully they'll be useful to any other built-in xblocks we create.
* js tests: 	de46f04 removes the `MathJax` cleanup performed by the `DiscussionThreadProfileView` tests, to allow the [loaded MathJax library](https://github.com/edx/edx-platform/pull/13483/files) to continue to work for subsequent tests.

If possible, can you finish your review before Monday noon EDT?  We're trying to avoid another week's spillover.  Thank you for your help!

pomegranited commented:
Sweet, thank you @andy-armstrong !  Do I need any more ok's before squashing and merging?

pomegranited commented:
@douglashall I noticed you're in the Reviewers list for this PR.   Did you need anything changed before we squash & merge?

douglashall commented:
@pomegranited Looks like this PR has already been thoroughly reviewed. I think you are good to merge. Thanks!

andy-armstrong commented:
@pomegranited It's all set from my standpoint, since both you and I have approved the changes. I'd go ahead and squash so it is ready to merge, since @douglashall won't need to see the individual commits.

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
jenkins run lettuce

andy-armstrong commented:
@pomegranited @itsjeyd @bradenmacdonald this is all set to merge (the coveralls failure is a known random issue that is safe to ignore).

pomegranited commented:
Thanks @andy-armstrong , @douglashall !  Merging now.

jzoldak commented:
@itsjeyd @pomegranited @andy-armstrong is the reason for the "jenkins run lettuce" comments that the "LMS.Answer problems" lettuce tests were failing or flaky? I'm wondering because they have become flaky on master since around the time that this PR was merged yesterday, and I'm trying to identify the root cause.

pomegranited commented:
Sorry, @jzoldak , I can't remember which tests were failing.  @andy-armstrong , do you know?

But it could be this PR.. Do those tests have anything to do with inline discussions or the Discussion tab?

andy-armstrong commented:
@jzoldak Here's a link to the flaky test:

https://build.testeng.edx.org/job/edx-platform-lettuce-pr/26140/

andy-armstrong commented:
It didn't seem to have anything to do with these changes so I figured it was okay to merge.

jzoldak commented:
@andy-armstrong yeah, my guess is a timing issue because the page now loads either faster or slower than it used to. 
Thx for the links, I'll dig in a bit and let you all know if I can gather more details.

jzoldak commented:
@andy-armstrong this is still an issue. The root cause of the lettuce test failure seems to be a method that navigates to a problem via URL but then does not wait for the page to fully render before interacting with it. 
I have an exploratory PR in but my first attempt at a fix didn't solve it. :( #13646 

Should we revert this, as it's affecting master and PR builds?
cc: @cahrens @benpatterson

andy-armstrong commented:
@jzoldak I don't want to revert this change as it seems unrelated to me, and this work unlocks a lot of future discussion stories. Can we just mark the test as flaky and/or disable it until we work out how to make it more reliable?

andy-armstrong commented:
@itsjeyd @pomegranited I'm afraid we're going to have to revert this commit as it has caused major instability in our Lettuce tests. We're still trying to determine what the issue is, but we are losing too much time to test build failures. See @jzoldak's revert PR for details: https://github.com/edx/edx-platform/issues/13651

There is some possibility that this is a MathJax issue since some of the failures are in previewing math, and this code did change the way that MathJax is loaded for discussions. It is possible that this in turn has affected other xmodules. The failures are not consistent, though, so this is only a guess at this point.

FYI @cahrens

pomegranited commented:
@andy-armstrong No worries, that's how it goes!  We'll assess the timing issues and re-submit.

CC @itsjeyd @antoviaque

cahrens commented:
With these changes, I was also seeing some different MathJax rendering in the preview area under inputs in problems. The input would shrink in height before the MathJax rendered, and it seemed slower to render (sometimes it would never render). It's possible this is just a coincidence, but it's something to test when this change is ready to go back in.
